CodeEditor: Add `->language()`, improve dark mode, add `destroy()` for cleanup

binaryfire opened the pull request:
CodeMirror only handles a single language at a time, so the current implementation wasn't working properly. This PR adds a `->language()` method which allows the user to select the language they want to use. By default it's set to plain text (none). 

**_Examples_**

Default (plain text):
```PHP
CodeEditor::make('code')
```
![image](https://github.com/user-attachments/assets/bfd530f7-aa1e-4286-86f0-e5dfef7547dc)

PHP:
```PHP
CodeEditor::make('code')
    ->language(Language::PHP),
```
![image](https://github.com/user-attachments/assets/bea5b674-550f-41eb-aa73-280ace08babb)

JSON:
```PHP
CodeEditor::make('code')
    ->language(Language::JSON),
```
![image](https://github.com/user-attachments/assets/e3535b9f-c902-4d1a-b2fb-3b40b16db66d)

binaryfire committed:
Add code editor language enum

binaryfire committed:
CodeEditor language getter and setter

binaryfire committed:
Add language var to code-editor Blade component

binaryfire committed:
Add language selection to code-editor.js

binaryfire committed:
Build new js

binaryfire committed:
Add language selection to code-editor.js

binaryfire committed:
Build new js

binaryfire committed:
Build code-editor.js

binaryfire committed:
Add PlainText language and make default

binaryfire committed:
Update CodeEditor docs

binaryfire committed:
Move enum into CodeEditor namespace

binaryfire committed:
Move enum into CodeEditor namespace

andrefelipe18 commented on packages/forms/resources/js/components/code-editor.js:
```diff
@@ -8,12 +8,32 @@ import { php } from '@codemirror/lang-php'
-
         state,
 
+        getLanguageExtension() {
```
I think this makes it easier to add new languages

```javascript
getLanguageExtension() {
    const extensions = {
        css: css,
        html: html,
        javascript: javascript,
        json: json,
        php: php
    };

    return (extensions[language]?.()) || [];
},
```

binaryfire replied:
@andrefelipe18 Good call. Updated.

andrefelipe18 replied:
I'm happy to contribute, let's hope PR passes

ralphjsmit replied:
Is there an extension possible for Blade?

binaryfire replied:
@ralphjsmit I looked into it and it didn't seem straightforward since languages need to be added at build time. CodeMirror doesn't have a plugin system or registry.

I'm maintaining my own fork which adds shell (via `codemirror/legacy-modes` + `StreamLanguage`) and Python. It increases bundle size though so not sure if Dan wants to open that door. The risk is people start requesting additional languages and the bundle size blows out.

There are 25 independent languages: https://github.com/orgs/codemirror/repositories?q=lang.
Plus all the legacy modes: https://github.com/codemirror/legacy-modes?tab=readme-ov-file#api-reference

I could add a few more core languages though. Shell, Python, SQL and YAML seem like good ones to me. That wouldn't add too much weight and would cover most use cases.

Someone with more JS experience might be able to figure out a way to dynamically add languages but that should be a separate PR. I'd like to just get the fundamentals working with this one.

ralphjsmit replied:
Okay, yes I suppose that's a good consideration. Blade I don't see indeed as a plugin, probably too niche. SQL you could consider, then you capture pretty much all technologies used around Laravel.

binaryfire replied:
@ralphjsmit Ah sorry, I misunderstood your question! I thought you were asking about dynamically adding more CodeMirror languages via PHP.

For Blade highlighting you’d need to write a custom language: https://codemirror.net/examples/lang-package/. Definitely doable but it’d take some work and would need to be maintained. So not something that should go in core.

ralphjsmit replied:
Ah yeah okay, yeah just Blade-language support :) But if it requires a new plugin then that's not really an option for Filament.

binaryfire committed:
Improve code-editor.js

ralphjsmit commented on packages/forms/src/Components/CodeEditor/Enums/Language.php:
```diff
@@ -0,0 +1,18 @@
+
+    case PHP = 'php';
+
+    case PlainText = 'plaintext';
```
If it is "PlainText", should the key be `plain_text`?

```suggestion
    case PlainText = 'plain_text';
```

binaryfire replied:
@ralphjsmit Yeah it should have been. Updated.

ralphjsmit reviewed (commented):
Nice addition! I missed this already when testing out the `CodeEditor` :)

binaryfire committed:
Update Language enum

binaryfire committed:
Improve dark mode, add destroy for proper component cleanup

binaryfire commented:
@danharrin I just added the OneDark theme (https://github.com/codemirror/theme-one-dark) to improve dark mode. OneDark is a first-party CodeMirror theme (actually it's their only theme) so there won't be any maintenance headaches. The default dark mode is _really_ bad.

Before:
![image](https://github.com/user-attachments/assets/003337f6-7586-484f-8dc0-fd9662606f63)

After:
![image](https://github.com/user-attachments/assets/c9ba805c-5676-4caa-a837-785a536c5931)

I've also added a destroy method to ensure proper component cleanup.

binaryfire committed:
Fix dark mode rounded corners

binaryfire committed:
Better rounded corners with overflow-hidden (fixes scrollbar corners too)

binaryfire committed:
Fix CSS

danharrin commented on packages/forms/resources/css/components/code-editor.css:
```diff
@@ -1,11 +1,13 @@
 
         & .cm-gutters {
-            @apply min-h-48! rounded-s-lg border-e-gray-300! bg-gray-100! dark:border-e-gray-800! dark:bg-gray-950!;
+            @apply min-h-48! border-e-gray-300! bg-gray-100! dark:border-e-gray-800! dark:bg-gray-950!;
```
Can you please explain why the `rounded-s-lg` was removed?

binaryfire replied:
@danharrin Because adding `overflow-hidden` rounds the corners:

![image](https://github.com/user-attachments/assets/34473bc7-6bf7-4f7a-baf0-44a01cdcf5a1)

![image](https://github.com/user-attachments/assets/3fd890a1-4c85-4bca-bc0f-cfac1853382c)

And I thought the bottom staying square looked better when the scrollbar appears:

![image](https://github.com/user-attachments/assets/d3cf0e11-3ea7-4269-8045-203630e61914)

(vs this):

![image](https://github.com/user-attachments/assets/c4e8d1e6-25d6-498e-81ab-03ea93536648)

Up to you though - I can re-add it if you prefer.

danharrin commented on packages/forms/docs/20-code-editor.md:
````diff
@@ -14,3 +14,15 @@ CodeEditor::make('code')
+
+## Changing the code editor's language
+
+You may change the language of the code editor using the `language()` method. The editor supports CSS, HTML, JavaScript, JSON, PHP and plain text. You can open the `Filament\Forms\Components\CodeEditor\Enums\Language` enum class to see the full list. To switch to use JavaScript as the language, you can use the `Language::Javascript` enum value:
+
+```php
+use Filament\Forms\Components\CodeEditor;
+use Filament\Forms\Components\CodeEditor\Enums\Language;
+
+CodeEditor::make('code')
+    ->language(Language::Javascript)
````
```suggestion
You may change the language of the code editor using the `language()` method. The editor supports CSS, HTML, JavaScript, JSON, PHP and plain text. You can open the `Filament\Forms\Components\CodeEditor\Enums\Language` enum class to see the full list. To switch to using JavaScript as the language, you can use the `Language::JavaScript` enum value:

```php
use Filament\Forms\Components\CodeEditor;
use Filament\Forms\Components\CodeEditor\Enums\Language;

CodeEditor::make('code')
    ->language(Language::JavaScript)
```

danharrin commented on packages/forms/resources/js/components/code-editor.js:
```diff
@@ -5,15 +5,40 @@ import { html } from '@codemirror/lang-html'
+        themeCompartment: new Compartment(),
         state,
 
+        isDarkMode() {
```
```suggestion
        isDarkMode: function () {
```

danharrin commented on packages/forms/resources/js/components/code-editor.js:
```diff
@@ -5,15 +5,40 @@ import { html } from '@codemirror/lang-html'
+            return document.documentElement.classList.contains('dark')
+        },
+
+        getThemeExtensions() {
```
```suggestion
        getThemeExtensions: function () {
```

danharrin commented on packages/forms/resources/js/components/code-editor.js:
```diff
@@ -5,15 +5,40 @@ import { html } from '@codemirror/lang-html'
+            return this.isDarkMode() ? [oneDark] : []
+        },
+
+        getLanguageExtension() {
```
```suggestion
        getLanguageExtension: function () {
```

danharrin commented on packages/forms/resources/js/components/code-editor.js:
```diff
@@ -5,15 +5,40 @@ import { html } from '@codemirror/lang-html'
+
+        getLanguageExtension() {
+            const extensions = {
+                css: css,
+                html: html,
+                javascript: javascript,
+                json: json,
+                php: php,
```
```suggestion
                css,
                html,
                javascript,
                json,
                php,
```

danharrin commented on packages/forms/src/Components/CodeEditor.php:
```diff
@@ -12,4 +13,18 @@ class CodeEditor extends Field
      */
     protected string $view = 'filament-forms::components.code-editor';
+
+    protected Language $language = Language::PlainText;
+
+    public function language(Language $language): static
+    {
+        $this->language = $language;
+
+        return $this;
+    }
+
+    public function getLanguage(): Language
+    {
+        return $this->language;
+    }
```
Please allow closure functions to be passed and evaluated

Ensure that the `<UtilityInjection>` component is used in the docs to reflect this

danharrin commented on packages/forms/src/Components/CodeEditor/Enums/Language.php:
```diff
@@ -0,0 +1,18 @@
+
+enum Language: string
+{
+    case CSS = 'css';
+
+    case HTML = 'html';
+
+    case Javascript = 'javascript';
+
+    case JSON = 'json';
+
+    case PHP = 'php';
```
```suggestion
    case Css = 'css';

    case Html = 'html';

    case JavaScript = 'javascript';

    case Json = 'json';

    case Php = 'php';
```

binaryfire committed:
Refactor code-editor.js

binaryfire committed:
Update Language enum values

binaryfire committed:
Support closures in language() and update docs

binaryfire commented:
@danharrin I've made all the changes except the CSS: https://github.com/filamentphp/filament/pull/16407#discussion_r2133638695

danharrin commented on packages/forms/docs/20-code-editor.md:
````diff
@@ -14,3 +14,17 @@ CodeEditor::make('code')
+    ->language(Language::JavaScript)
+```
+
+<UtilityInjection set="formFields" version="4.x">As well as allowing a static value, the language() method also accepts a function to dynamically calculate it. You can inject various utilities into the function as parameters.</UtilityInjection>
````
This component needs to be imported at the top

danharrin commented on packages/forms/src/Components/CodeEditor.php:
```diff
@@ -12,4 +14,18 @@ class CodeEditor extends Field
      */
     protected string $view = 'filament-forms::components.code-editor';
+
+    protected Language | Closure $language = Language::PlainText;
```
Instead of needing a `PlainText` case, I think that should just be represented by `null` language.

danharrin commented on packages/forms/resources/js/components/code-editor.js:
```diff
@@ -5,15 +5,40 @@ import { html } from '@codemirror/lang-html'
+                php,
+            }
+
+            return extensions[language]?.() || []
```
Instead of returning an extension or an array, can we return an extension or null, and then when using the extension only register it if its not null?

binaryfire replied:
I think the current implementation is cleaner and more performant:

Current approach (returning []):
```JS
extensions: [
    basicSetup,
    keymap.of([indentWithTab]),
    // ... other extensions
    this.getLanguageExtension(), // Clean - returns [] when no lang set
    this.themeCompartment.of(this.getThemeExtensions()),
]
```

And we're only calling `getLanguageExtension()` once

Proposed approach (returning null) would require:
```JS
// Option 1: Calls getLanguageExtension() twice
...(this.getLanguageExtension() ? [this.getLanguageExtension()] : []),

// Option 2: More verbose
const langExt = this.getLanguageExtension();
if (langExt) extensions.push(langExt);
```

The language extension is already only being registered if set. If not set, `getLanguageExtension()` returns an empty array (i.e. no extension).

danharrin replied:
I just don't like the inconsistency where an empty array represents no extension instead of `null`. It feels like we are just relying on the fact that CodeMirror appears to ignore empty arrays in the extension list by chance, which feels fragile to assume.

A constant can be used to store the resolved extension object so the method only needs to be run once

danharrin commented on packages/forms/docs/20-code-editor.md:
````diff
@@ -14,3 +14,17 @@ CodeEditor::make('code')
 ```
 
 <AutoScreenshot name="forms/fields/code-editor/simple" alt="Code editor" version="4.x" />
````
Now that there isn't syntax highlighting out of the box, the code used to generate this screenshot in `FieldsDemo.php` needs to be updated to ensure it has the correct language

danharrin commented on packages/forms/docs/20-code-editor.md:
The "Introduction" section shouldn't mention the available languages anymore, as this implies they can be used without additional config, especially since the first code example doesn't use them.

binaryfire committed:
Remove PlainText language, use null instead

binaryfire committed:
Update CodeEditor docs

binaryfire commented:
@danharrin Done. 

- Removed `PlainText` case, use `null` by default instead
- Updated docs, added econd screenshot to demonstrate syntax highlighting.
- The current code already only registers extensions when they exist, since `getLanguageExtension()` returns an empty array when no lang is set: https://github.com/filamentphp/filament/pull/16407#discussion_r2134384798

binaryfire committed:
Update CodeEditor docs

danharrin committed:
compile screenshots

danharrin committed:
make requested changes

danharrin committed:
Update language.jpg

danharrin committed:
Update 20-code-editor.md

danharrin committed:
update screenshots

danharrin committed:
Merge branch '4.x' into pr/16407
